### PR TITLE
Update dashboard to v2021.12.31.1

### PR DESCRIPTION
**Pull request by baurine**
Signed-off-by: baurine <2008.hbl@gmail.com>

### What problem does this PR solve?

Ref #4257 

Update TiDB Dashboard to v2021.11.08.1.

Upstream commit: https://github.com/pingcap/tidb-dashboard/commit/b2c4a0435690c75bfb09117e8f52908552f9abec .

Related tidb-dashboard PRs:

- https://github.com/pingcap/tidb-dashboard/pull/1113
- https://github.com/pingcap/tidb-dashboard/pull/1120

### Release note

```release-note
1. Refine manual profiling and continuous profiling UI
2. Use speedscope to show flamegraph
3. Support multiple profiling types for manual profiling
4. Support dynamic distro res
5. New feature: Top SQL
```

### CODE REVIEW

File: pkg/dashboard/adapter/config.go [hunk 1]
```diff
@@ -35,7 +35,6 @@ func GenDashboardConfig(srv *server.Server) (*config.Config, error) {
 	dashboardCfg.PublicPathPrefix = cfg.Dashboard.PublicPathPrefix
 	dashboardCfg.EnableTelemetry = cfg.Dashboard.EnableTelemetry
 	dashboardCfg.EnableExperimental = cfg.Dashboard.EnableExperimental
-	dashboardCfg.EnableNonRootLogin = true
```
Comment (breeswish) [inline pkg/dashboard/adapter/config.go:38]: Why?

Comment (baurine) [reply pkg/dashboard/adapter/config.go:38]: The `EnableNonRootLogin` config has been abandoned in the dashboard side after using the feature flag strategy.

File: pkg/dashboard/uiserver/embedded_assets_rewriter.go [hunk 2]
```diff
@@ -17,19 +17,30 @@ package uiserver
...
 func Assets(cfg *config.Config) http.FileSystem {
 	once.Do(func() {
-		uiserver.RewriteAssets(assets, cfg, func(fs http.FileSystem, f http.File, path, newContent string, bs []byte) {
+
+		exePath, err := os.Executable()
+		if err != nil {
+			log.Fatal("Failed to get executable path", zap.Error(err))
+		}
+
+		distroResFolderPath := path.Join(path.Dir(exePath), "distro-res")
```
Comment (breeswish) [inline pkg/dashboard/uiserver/embedded_assets_rewriter.go:41]: These logics are duplicated with `dashboard.go`. Could you have some refinements?

Comment (baurine) [reply pkg/dashboard/uiserver/embedded_assets_rewriter.go]: done!

File: pkg/dashboard/distro/distro.go [hunk 3]
```diff
@@ -12,15 +12,46 @@
...
+const (
+	distroResFolderName      string = "distro-res"
+	distroStringsResFileName string = "strings.json"
+)
+
+func GetDistroResFolderPath() (string, error) {
+	exePath, err := os.Executable()
+	if err != nil {
+		return "", err
+	}
```
Comment (baurine) [inline pkg/dashboard/distro/distro.go:39]: Is it ok to `log.Fatal()` inside this method?

**Comment by breeswish** [comment]
CI is failed. Please fix the CI~

**Comment by breeswish** [comment]
@baurine CI is not working. Please fix the CI, then find some one to approve it in tomorrow morning. Otherwise it won't catch up the code freeze.

**Review by breeswish — COMMENTED** [review]
There are also some lint failures.

File: .github/workflows/check-dashboard.yaml [hunk 4]
```diff
@@ -1,53 +0,0 @@
...
-        run: |
-          rm -rf /tmp/dashboard
-          mkdir -p /tmp/dashboard
-          cd /tmp/dashboard
-          DOWNLOAD_URL="https://github.com/pingcap/tidb-dashboard/archive/v${DASHBOARD_RELEASE_VERSION}.zip"
-          curl -L "${DOWNLOAD_URL}" --fail --output source.zip
-          unzip source.zip
-      - name: Check diff
-        run: |
-          diff -qr ${DASHBOARD_DIR}/ui/ /tmp/dashboard/tidb-dashboard-${DASHBOARD_RELEASE_VERSION}/ui/
```
Comment (baurine) [inline .github/workflows/check-dashboard.yaml:53]: `${DASHBOARD_DIR}` doesn't download ui folder anymore, so we remove this check.

File: go.mod [hunk 5]
```diff
@@ -50,8 +50,9 @@ require (
 	go.etcd.io/bbolt v1.3.5 // indirect
 	go.etcd.io/etcd v0.5.0-alpha.5.0.20191023171146-3cf2f69b5738
 	go.uber.org/goleak v1.1.12
-	go.uber.org/zap v1.16.0
+	go.uber.org/zap v1.19.0
 	golang.org/x/tools v0.1.5
+	google.golang.org/appengine v1.6.5 // indirect
```
[the remark this answers is not in the document]
Comment (baurine) [reply go.mod:55]: Yep, let me try.

**Comment by breeswish** [comment]
I'm manually verifying these cases, will update the state after finished:

- [ ] Use a release version
- [ ] Use a master version
- [ ] Specify distro for the compiled binary

**Review by breeswish — APPROVED** [review]

**Comment by baurine** [comment]
wait a new release: https://github.com/pingcap/tidb-dashboard/pull/1120

**Review by breeswish — APPROVED** [review]
Manually tested.

**Review by HunDunDM — APPROVED** [review]

**Review by lhy1024 — APPROVED** [review]

**Pull request merged**